File: src/main/scala/com/abaddon83/cardsGames/burracoGames/commands/AppendCardOnBurracoCmd.scala
package com.abaddon83.cardsGames.burracoGames.commands

import java.util.UUID

import com.abaddon83.cardsGames.burracoGames.domainModels.BurracoId
import com.abaddon83.cardsGames.burracoGames.ports.BurracoGameRepositoryPort
import com.abaddon83.cardsGames.shares.decks.Card
import com.abaddon83.cardsGames.shares.games.GameIdentity
import com.abaddon83.cardsGames.shares.players.PlayerIdentity
import com.abaddon83.libs.cqs.commands.{Command, CommandHandler}

import scala.concurrent.{Await, Future}

case class AppendCardOnBurracoCmd(
                                gameIdentity: GameIdentity,
                                playerIdentity: PlayerIdentity,
                                burracoId: BurracoId,
                                cardsToAppend: List[Card]
                              ) extends Command {
  override protected val requestId: UUID = UUID.randomUUID()
}

class AppendCardOnBurracoHandler(
                                  gameRepositoryPort: BurracoGameRepositoryPort
                                )
                                (implicit val ec: scala.concurrent.ExecutionContext)
  extends CommandHandler[AppendCardOnBurracoCmd] {

  override def handleAsync(command: AppendCardOnBurracoCmd): Future[Unit] = {
    for {
      gameExecution <- gameRepositoryPort.findBurracoGameInitialisedTurnExecutionBy(gameIdentity = command.gameIdentity)
      gameExecutionUpdated = gameExecution.appendCardsOnABurracoDropped(playerIdentity = command.playerIdentity,cardsToAppend = command.cardsToAppend,burracoId = command.burracoId )
    } yield gameRepositoryPort.save(gameExecutionUpdated)
  }

  override def handle(command: AppendCardOnBurracoCmd): Unit = {
    import scala.concurrent.duration._
    import scala.language.postfixOps
    val gameExecution = Await.result(gameRepositoryPort.findBurracoGameInitialisedTurnExecutionBy(gameIdentity = command.gameIdentity), 500 millis)
    val gameExecutionUpdated = gameExecution.appendCardsOnABurracoDropped(playerIdentity = command.playerIdentity,cardsToAppend = command.cardsToAppend,burracoId = command.burracoId )
    gameRepositoryPort.save(gameExecutionUpdated)
  }

}

File: src/main/kotlin/com/abaddon83/burracoGame/commands/AppendCardOnBurracoCmd.kt
package com.abaddon83.burracoGame.commands

import com.abaddon83.burracoGame.ports.BurracoGameRepositoryPort
import com.abaddon83.burracoGame.shared.burracos.BurracoIdentity
import com.abaddon83.burracoGame.shared.decks.Card
import com.abaddon83.burracoGame.shared.games.GameIdentity
import com.abaddon83.burracoGame.shared.players.PlayerIdentity
import com.abaddon83.utils.cqs.commands.Command
import com.abaddon83.utils.cqs.commands.CommandHandler
import kotlinx.coroutines.runBlocking
import org.koin.core.KoinComponent
import org.koin.core.inject
import java.sql.Timestamp
import java.util.*

data class AppendCardOnBurracoCmd(
        val gameIdentity: GameIdentity,
        val playerIdentity: PlayerIdentity,
        val burracoIdentity: BurracoIdentity,
        val cardsToAppend: List<Card>,
        override val commandId: UUID,
        override val commandTimeStamp: Timestamp
) : Command {
    constructor(gameIdentity: GameIdentity, playerIdentity: PlayerIdentity, burracoIdentity: BurracoIdentity, cardsToAppend: List<Card>) :
            this(
                    gameIdentity = gameIdentity,
                    playerIdentity = playerIdentity,
                    burracoIdentity = burracoIdentity,
                    cardsToAppend = cardsToAppend,
                    commandId = UUID.randomUUID(),
                    commandTimeStamp = Timestamp(System.currentTimeMillis()
                    )
            )
}

class AppendCardOnBurracoHandler() : CommandHandler<AppendCardOnBurracoCmd>, KoinComponent {

    private val repository: BurracoGameRepositoryPort by inject()


    override fun handle(command: AppendCardOnBurracoCmd) {
        val gameExecution = runBlocking {repository.findBurracoGameExecutionTurnExecutionBy(gameIdentity = command.gameIdentity)}
        check(gameExecution != null) {"GameIdentity ${command.gameIdentity} not found exist"}
        val gameExecutionUpdated = gameExecution.appendCardsOnABurracoDropped(
                playerIdentity = command.playerIdentity,
                cardsToAppend = command.cardsToAppend,
                burracoIdentity = command.burracoIdentity )
        repository.save(gameExecutionUpdated)
    }

    override suspend fun handleAsync(command: AppendCardOnBurracoCmd) {
        val gameExecution = repository.findBurracoGameExecutionTurnExecutionBy(gameIdentity = command.gameIdentity)
        check(gameExecution != null) {"GameIdentity ${command.gameIdentity} not found exist"}
        val gameExecutionUpdated = gameExecution.appendCardsOnABurracoDropped(
                playerIdentity = command.playerIdentity,
                cardsToAppend = command.cardsToAppend,
                burracoIdentity = command.burracoIdentity )
        repository.save(gameExecutionUpdated)
    }
}